examples/step-58/step-58.cc

```diff
@@ -0,0 +1,673 @@
+  // read through the tutorial programs at least up to step-6:
+  template <int dim>
+  void NonlinearSchroedingerEquation<dim>::setup_system()
+  {
+    GridGenerator::hyper_cube(triangulation, -1, 1);
+    triangulation.refine_global(6);
+
+    std::cout << "Number of active cells: " << triangulation.n_active_cells()
+              << std::endl;
+
+    dof_handler.distribute_dofs(fe);
+
+    std::cout << "Number of degrees of freedom: " << dof_handler.n_dofs()
+              << std::endl
+              << std::endl;
+
+    DynamicSparsityPattern dsp(dof_handler.n_dofs(), dof_handler.n_dofs());
+    DoFTools::make_sparsity_pattern(dof_handler, dsp);
+    sparsity_pattern.copy_from(dsp);
+
+    system_matrix.reinit(sparsity_pattern);
+    rhs_matrix.reinit(sparsity_pattern);
+
+    solution.reinit(dof_handler.n_dofs());
+    system_rhs.reinit(dof_handler.n_dofs());
+
+    constraints.close();
+  }
+
+
+
+  // Next, we assemble the relevant matrices. The way we have written
+  // the Crank-Nicolson discretization of the spatial step of the Strang
+  // splitting (i.e., the second of the three partial steps in each time
+  // step), we were led to the linear system
+  // $\left[ -iM  +  \frac 14 k_{n+1} A + \frac 12 k_{n+1} W \right]
+  //   \Psi^{(n,2)}
+  //  =
+  //  \left[ -iM  -  \frac 14 k_{n+1} A - \frac 12 k_{n+1} W \right]
+  //   \Psi^{(n,1)}$.
+  // In other words, there are two matrices in play here -- one for the
+  // left and one for the right hand side. We build these matrices
+  // separately. (One could avoid building the right hand side matrix
+  // and instead just form the *action* of the matrix on $\Psi^{(n,1)}$
+  // in each time step. This may or may not be more efficient, but
+  // efficiency is not foremost on our minds for this program.)
+  template <int dim>
+  void NonlinearSchroedingerEquation<dim>::assemble_matrices()
+  {
+    const QGauss<dim> quadrature_formula(fe.degree + 1);
+
+    FEValues<dim> fe_values(fe,
+                            quadrature_formula,
+                            update_values | update_gradients |
+                              update_quadrature_points | update_JxW_values);
+
+    const unsigned int dofs_per_cell = fe.dofs_per_cell;
+    const unsigned int n_q_points    = quadrature_formula.size();
+
+    FullMatrix<std::complex<double>> cell_matrix_lhs(dofs_per_cell,
+                                                     dofs_per_cell);
+    FullMatrix<std::complex<double>> cell_matrix_rhs(dofs_per_cell,
+                                                     dofs_per_cell);
+
+    std::vector<types::global_dof_index> local_dof_indices(dofs_per_cell);
+    std::vector<double>                  potential_values(n_q_points);
+    const Potential<dim>                 potential;
+
+    for (const auto &cell : dof_handler.active_cell_iterators())
+      {
+        cell_matrix_lhs = std::complex<double>(0.);
+        cell_matrix_rhs = std::complex<double>(0.);
```

Comment:
I'm kind of surprised that we don't do type promotion for the RHS. In the docs I believe (although I haven't double checked) that we write something about `cell_matrix = 0` being *the* way to reset a matrix. Maybe we need to check/update the docs, or add some `operator=(int)` to `FullMatrix` that does this internally.

Reply:
Yes, adding `operator=(int)` would resolve this issue. The problem is that there is no automatic conversion of either `double` to `std::complex<double>`, not `int` to `std::complex<double>`. I think that's actually a mistake in the standard, but that's outside my realm of action.